### ISSUE TITLE
Not the Same Performance At 'CNN-non-static'

### ISSUE THREAD

**Issue by 3rduncle** [alexander-rakhlin/CNN-for-Sentence-Classification-in-Keras#1]
Hi, I follow your setting below about 'CNN-non-static' network.
```py
model_variation = 'CNN-non-static'  #  CNN-rand | CNN-non-static | CNN-static
print('Model variation is %s' % model_variation)

# Model Hyperparameters
sequence_length = 56
embedding_dim = 20
filter_sizes = (3, 4)
num_filters = 3
dropout_prob = (0.7, 0.8)
hidden_dims = 100
```

Then I split 10% samples for test. 
```py
# Shuffle data
shuffle_indices = np.random.permutation(np.arange(len(y)))
x_shuffled = x[shuffle_indices]
y_shuffled = y[shuffle_indices].argmax(axis=1)

test_split = 0.1
split_at = len(y) * (1 - test_split)
x_train, y_train = x_shuffled[:split_at], y_shuffled[:split_at]
x_val, y_val = x_shuffled[split_at:], y_shuffled[split_at:]
```

The network trains with the other 90% samples under the setting   ``` validation_split=0.1```. The test samples will be evaluated after the last epoch.
```py
model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

# Training model
# ==================================================
model.fit(x_train, y_train, batch_size=batch_size, nb_epoch=num_epochs, validation_split=val_split, verbose=2)
e = model.evaluate(x_val, y_val, batch_size=len(y_val), verbose=2)
print('loss %f acc %f' % tuple(e))
```
The trainning log seems perfect. the val_acc goes beyond 90%, but the test performance is sad.
It is only 76% far away from your result.
It looks like the validation is used for training over epochs, leading to the good performance of validation.

Is there some mistakes I made?
```py
3s - loss: 0.2998 - acc: 0.8870 - val_loss: 0.1997 - val_acc: 0.9344
Epoch 91/100
3s - loss: 0.2971 - acc: 0.8923 - val_loss: 0.2010 - val_acc: 0.9302
Epoch 92/100
3s - loss: 0.2856 - acc: 0.8935 - val_loss: 0.1979 - val_acc: 0.9250
Epoch 93/100
3s - loss: 0.2831 - acc: 0.8940 - val_loss: 0.1994 - val_acc: 0.9302
Epoch 94/100
3s - loss: 0.2818 - acc: 0.8903 - val_loss: 0.1967 - val_acc: 0.9333
Epoch 95/100
3s - loss: 0.2895 - acc: 0.8892 - val_loss: 0.2103 - val_acc: 0.9187
Epoch 96/100
3s - loss: 0.2774 - acc: 0.8953 - val_loss: 0.2021 - val_acc: 0.9292
Epoch 97/100
3s - loss: 0.2739 - acc: 0.8957 - val_loss: 0.2453 - val_acc: 0.9073
Epoch 98/100
3s - loss: 0.2771 - acc: 0.8945 - val_loss: 0.1993 - val_acc: 0.9198
Epoch 99/100
3s - loss: 0.2798 - acc: 0.8960 - val_loss: 0.1953 - val_acc: 0.9250
Epoch 100/100
3s - loss: 0.2810 - acc: 0.8932 - val_loss: 0.2059 - val_acc: 0.9167
loss 0.635195 acc 0.761012
```

**Comment by alexander-rakhlin** [comment]
Hi!

Thanks for the interest. I run your setup for 30 epochs with np.random.seed(2). Haven't noticed anything unfamiliar:

`Epoch 27/30
12s - loss: 0.3765 - acc: 0.8383 - val_loss: 0.4473 - val_acc: 0.7958
Epoch 28/30
11s - loss: 0.3807 - acc: 0.8398 - val_loss: 0.4496 - val_acc: 0.7917
Epoch 29/30
11s - loss: 0.3724 - acc: 0.8461 - val_loss: 0.4527 - val_acc: 0.7917
Epoch 30/30
loss 0.486175 acc 0.782568`

Looks like we use different Keras builds, but this shouldn't affect. Possible reason is a small data set, and you make it event smaller. val_acc: 0.9167 is highly unrealistic and suggests validation bias is large. This is possible, but I haven't seen this in my experiments yet. I can only suggest trying other random seeds.

**Comment by lan2720** [comment]
Hi, I run your code with the setting 'CNN-non-static'.
The log as follow. I wonder why I can't get your 82.1% result.
```
Using TensorFlow backend.
Model variation is CNN-non-static
Loading data...
Loading existing Word2Vec model '20features_1minwords_10context'
Vocabulary Size: 18765
Train on 8635 samples, validate on 960 samples
Epoch 1/100
4s - loss: 0.6939 - acc: 0.5065 - val_loss: 0.6932 - val_acc: 0.5000
Epoch 2/100
4s - loss: 0.6936 - acc: 0.5018 - val_loss: 0.6925 - val_acc: 0.5646
Epoch 3/100
4s - loss: 0.6912 - acc: 0.5276 - val_loss: 0.6880 - val_acc: 0.5437
Epoch 4/100
4s - loss: 0.6817 - acc: 0.5647 - val_loss: 0.6711 - val_acc: 0.5875
Epoch 5/100
4s - loss: 0.6652 - acc: 0.5984 - val_loss: 0.6622 - val_acc: 0.5885
Epoch 6/100
4s - loss: 0.6457 - acc: 0.6343 - val_loss: 0.6305 - val_acc: 0.6219
Epoch 7/100
5s - loss: 0.6156 - acc: 0.6637 - val_loss: 0.6113 - val_acc: 0.6302
Epoch 8/100
5s - loss: 0.5939 - acc: 0.6899 - val_loss: 0.5687 - val_acc: 0.6958
Epoch 9/100
5s - loss: 0.5725 - acc: 0.7090 - val_loss: 0.5602 - val_acc: 0.6969
Epoch 10/100
5s - loss: 0.5488 - acc: 0.7303 - val_loss: 0.5753 - val_acc: 0.6937
Epoch 11/100
5s - loss: 0.5251 - acc: 0.7493 - val_loss: 0.5569 - val_acc: 0.7135
Epoch 12/100
5s - loss: 0.5043 - acc: 0.7617 - val_loss: 0.5253 - val_acc: 0.7448
Epoch 13/100
5s - loss: 0.4956 - acc: 0.7671 - val_loss: 0.5356 - val_acc: 0.7385
Epoch 14/100
5s - loss: 0.4872 - acc: 0.7776 - val_loss: 0.5576 - val_acc: 0.7260
Epoch 15/100
5s - loss: 0.4723 - acc: 0.7804 - val_loss: 0.5335 - val_acc: 0.7521
Epoch 16/100
5s - loss: 0.4603 - acc: 0.7933 - val_loss: 0.5194 - val_acc: 0.7594
Epoch 17/100
5s - loss: 0.4516 - acc: 0.7970 - val_loss: 0.5055 - val_acc: 0.7667
Epoch 18/100
5s - loss: 0.4481 - acc: 0.8014 - val_loss: 0.5084 - val_acc: 0.7677
Epoch 19/100
5s - loss: 0.4399 - acc: 0.8030 - val_loss: 0.5019 - val_acc: 0.7667
Epoch 20/100
5s - loss: 0.4259 - acc: 0.8131 - val_loss: 0.5127 - val_acc: 0.7667
Epoch 21/100
6s - loss: 0.4274 - acc: 0.8162 - val_loss: 0.5301 - val_acc: 0.7615
Epoch 22/100
5s - loss: 0.4035 - acc: 0.8286 - val_loss: 0.5048 - val_acc: 0.7698
Epoch 23/100
5s - loss: 0.4120 - acc: 0.8278 - val_loss: 0.5056 - val_acc: 0.7677
Epoch 24/100
5s - loss: 0.4041 - acc: 0.8241 - val_loss: 0.5010 - val_acc: 0.7708
Epoch 25/100
5s - loss: 0.4013 - acc: 0.8280 - val_loss: 0.5367 - val_acc: 0.7552
Epoch 26/100
6s - loss: 0.3985 - acc: 0.8339 - val_loss: 0.5084 - val_acc: 0.7667
Epoch 27/100
5s - loss: 0.3935 - acc: 0.8349 - val_loss: 0.5014 - val_acc: 0.7667
Epoch 28/100
6s - loss: 0.3881 - acc: 0.8368 - val_loss: 0.5076 - val_acc: 0.7677
Epoch 29/100
5s - loss: 0.3938 - acc: 0.8338 - val_loss: 0.4997 - val_acc: 0.7656
Epoch 30/100
5s - loss: 0.3832 - acc: 0.8401 - val_loss: 0.5031 - val_acc: 0.7688
Epoch 31/100
5s - loss: 0.3823 - acc: 0.8408 - val_loss: 0.5097 - val_acc: 0.7750
Epoch 32/100
5s - loss: 0.3742 - acc: 0.8457 - val_loss: 0.5153 - val_acc: 0.7667
Epoch 33/100
5s - loss: 0.3712 - acc: 0.8471 - val_loss: 0.5031 - val_acc: 0.7677
Epoch 34/100
5s - loss: 0.3705 - acc: 0.8478 - val_loss: 0.5023 - val_acc: 0.7604
Epoch 35/100
5s - loss: 0.3648 - acc: 0.8478 - val_loss: 0.5266 - val_acc: 0.7719
Epoch 36/100
5s - loss: 0.3643 - acc: 0.8544 - val_loss: 0.5102 - val_acc: 0.7573
Epoch 37/100
5s - loss: 0.3646 - acc: 0.8552 - val_loss: 0.5380 - val_acc: 0.7708
Epoch 38/100
5s - loss: 0.3666 - acc: 0.8533 - val_loss: 0.5115 - val_acc: 0.7604
Epoch 39/100
5s - loss: 0.3540 - acc: 0.8603 - val_loss: 0.5257 - val_acc: 0.7490
Epoch 40/100
5s - loss: 0.3577 - acc: 0.8558 - val_loss: 0.5177 - val_acc: 0.7573
Epoch 41/100
5s - loss: 0.3510 - acc: 0.8586 - val_loss: 0.5190 - val_acc: 0.7656
Epoch 42/100
5s - loss: 0.3521 - acc: 0.8543 - val_loss: 0.5426 - val_acc: 0.7656
Epoch 43/100
5s - loss: 0.3458 - acc: 0.8583 - val_loss: 0.5250 - val_acc: 0.7490
Epoch 44/100
5s - loss: 0.3534 - acc: 0.8594 - val_loss: 0.5260 - val_acc: 0.7667
Epoch 45/100
5s - loss: 0.3469 - acc: 0.8616 - val_loss: 0.5305 - val_acc: 0.7552
Epoch 46/100
5s - loss: 0.3519 - acc: 0.8605 - val_loss: 0.5212 - val_acc: 0.7688
Epoch 47/100
5s - loss: 0.3484 - acc: 0.8649 - val_loss: 0.5291 - val_acc: 0.7625
Epoch 48/100
5s - loss: 0.3479 - acc: 0.8669 - val_loss: 0.5172 - val_acc: 0.7688
Epoch 49/100
5s - loss: 0.3393 - acc: 0.8611 - val_loss: 0.5306 - val_acc: 0.7677
Epoch 50/100
5s - loss: 0.3446 - acc: 0.8665 - val_loss: 0.5402 - val_acc: 0.7698
Epoch 51/100
5s - loss: 0.3331 - acc: 0.8683 - val_loss: 0.5246 - val_acc: 0.7615
Epoch 52/100
6s - loss: 0.3499 - acc: 0.8629 - val_loss: 0.5312 - val_acc: 0.7625
Epoch 53/100
7s - loss: 0.3304 - acc: 0.8686 - val_loss: 0.5597 - val_acc: 0.7667
Epoch 54/100
5s - loss: 0.3340 - acc: 0.8677 - val_loss: 0.5374 - val_acc: 0.7521
Epoch 55/100
5s - loss: 0.3256 - acc: 0.8733 - val_loss: 0.5528 - val_acc: 0.7646
Epoch 56/100
5s - loss: 0.3266 - acc: 0.8696 - val_loss: 0.5326 - val_acc: 0.7656
Epoch 57/100
5s - loss: 0.3314 - acc: 0.8724 - val_loss: 0.5416 - val_acc: 0.7698
Epoch 58/100
5s - loss: 0.3311 - acc: 0.8726 - val_loss: 0.5616 - val_acc: 0.7688
Epoch 59/100
5s - loss: 0.3286 - acc: 0.8730 - val_loss: 0.5196 - val_acc: 0.7646
Epoch 60/100
5s - loss: 0.3278 - acc: 0.8739 - val_loss: 0.5334 - val_acc: 0.7688
Epoch 61/100
5s - loss: 0.3365 - acc: 0.8728 - val_loss: 0.5354 - val_acc: 0.7667
Epoch 62/100
5s - loss: 0.3276 - acc: 0.8746 - val_loss: 0.5426 - val_acc: 0.7615
Epoch 63/100
5s - loss: 0.3205 - acc: 0.8786 - val_loss: 0.5743 - val_acc: 0.7677
Epoch 64/100
5s - loss: 0.3264 - acc: 0.8728 - val_loss: 0.5430 - val_acc: 0.7500
Epoch 65/100
5s - loss: 0.3140 - acc: 0.8806 - val_loss: 0.5556 - val_acc: 0.7656
Epoch 66/100
5s - loss: 0.3188 - acc: 0.8760 - val_loss: 0.5503 - val_acc: 0.7667
Epoch 67/100
5s - loss: 0.3272 - acc: 0.8726 - val_loss: 0.5343 - val_acc: 0.7698
Epoch 68/100
5s - loss: 0.3220 - acc: 0.8762 - val_loss: 0.5571 - val_acc: 0.7667
Epoch 69/100
5s - loss: 0.3211 - acc: 0.8779 - val_loss: 0.5303 - val_acc: 0.7646
Epoch 70/100
5s - loss: 0.3198 - acc: 0.8742 - val_loss: 0.5369 - val_acc: 0.7604
Epoch 71/100
5s - loss: 0.3122 - acc: 0.8818 - val_loss: 0.5521 - val_acc: 0.7688
Epoch 72/100
6s - loss: 0.3259 - acc: 0.8750 - val_loss: 0.5454 - val_acc: 0.7667
Epoch 73/100
6s - loss: 0.3103 - acc: 0.8840 - val_loss: 0.5401 - val_acc: 0.7594
Epoch 74/100
6s - loss: 0.3104 - acc: 0.8834 - val_loss: 0.5701 - val_acc: 0.7688
Epoch 75/100
5s - loss: 0.3222 - acc: 0.8792 - val_loss: 0.5415 - val_acc: 0.7531
Epoch 76/100
5s - loss: 0.3108 - acc: 0.8835 - val_loss: 0.5538 - val_acc: 0.7646
Epoch 77/100
5s - loss: 0.3154 - acc: 0.8830 - val_loss: 0.5354 - val_acc: 0.7573
Epoch 78/100
5s - loss: 0.3062 - acc: 0.8822 - val_loss: 0.5808 - val_acc: 0.7656
Epoch 79/100
5s - loss: 0.2970 - acc: 0.8835 - val_loss: 0.5915 - val_acc: 0.7573
Epoch 80/100
5s - loss: 0.3050 - acc: 0.8806 - val_loss: 0.5789 - val_acc: 0.7656
Epoch 81/100
5s - loss: 0.3041 - acc: 0.8873 - val_loss: 0.5336 - val_acc: 0.7635
Epoch 82/100
6s - loss: 0.3023 - acc: 0.8854 - val_loss: 0.5540 - val_acc: 0.7542
Epoch 83/100
5s - loss: 0.3041 - acc: 0.8847 - val_loss: 0.5748 - val_acc: 0.7625
Epoch 84/100
6s - loss: 0.3109 - acc: 0.8828 - val_loss: 0.6094 - val_acc: 0.7667
Epoch 85/100
6s - loss: 0.3074 - acc: 0.8866 - val_loss: 0.5559 - val_acc: 0.7750
Epoch 86/100
6s - loss: 0.3119 - acc: 0.8869 - val_loss: 0.5419 - val_acc: 0.7708
Epoch 87/100
5s - loss: 0.2985 - acc: 0.8849 - val_loss: 0.5771 - val_acc: 0.7781
Epoch 88/100
6s - loss: 0.2953 - acc: 0.8869 - val_loss: 0.5513 - val_acc: 0.7688
Epoch 89/100
5s - loss: 0.3064 - acc: 0.8835 - val_loss: 0.5612 - val_acc: 0.7625
Epoch 90/100
5s - loss: 0.2903 - acc: 0.8885 - val_loss: 0.5649 - val_acc: 0.7719
Epoch 91/100
5s - loss: 0.3048 - acc: 0.8823 - val_loss: 0.5724 - val_acc: 0.7698
Epoch 92/100
5s - loss: 0.3055 - acc: 0.8834 - val_loss: 0.5588 - val_acc: 0.7719
Epoch 93/100
5s - loss: 0.2984 - acc: 0.8903 - val_loss: 0.6106 - val_acc: 0.7708
Epoch 94/100
5s - loss: 0.3017 - acc: 0.8829 - val_loss: 0.5527 - val_acc: 0.7531
Epoch 95/100
5s - loss: 0.3021 - acc: 0.8851 - val_loss: 0.5639 - val_acc: 0.7771
Epoch 96/100
5s - loss: 0.2998 - acc: 0.8913 - val_loss: 0.5496 - val_acc: 0.7646
Epoch 97/100
5s - loss: 0.2942 - acc: 0.8879 - val_loss: 0.5679 - val_acc: 0.7688
Epoch 98/100
5s - loss: 0.2914 - acc: 0.8892 - val_loss: 0.5513 - val_acc: 0.7688
Epoch 99/100
5s - loss: 0.2985 - acc: 0.8870 - val_loss: 0.6031 - val_acc: 0.7771
Epoch 100/100
5s - loss: 0.2951 - acc: 0.8880 - val_loss: 0.5534 - val_acc: 0.7594
loss 0.531507 acc 0.777882

```
I have't changed the setting and just use
```
For 'CNN-non-static' gets to 82.1% after 61 epochs with following settings:
embedding_dim = 20          
filter_sizes = (3, 4)
num_filters = 3
dropout_prob = (0.7, 0.8)
hidden_dims = 100
```
Look forward to your reply. Thanks!: )

**Comment by alexander-rakhlin** [comment]
I haven't tried it with TF, and results depend on the net initialization. Besides, results vary due to relatively small data set. To summarize, your 76-78% look normal too.
